### PR TITLE
Create _header

### PR DESCRIPTION
**Pull request by marcel-wiedemeier**
Create _header file to specify security headers for netlify

### CODE REVIEW

**Comment by DaniRey** [comment]
Thank you for the contribution. I will merge it once the DNS alias is configured and HTTPS is configured.

Are you sure this file should be on top level? I would expect netlify to pick it up in the `docs` folder, because we configured `docs` as the base path for our project.

**Comment by marcel-wiedemeier** [comment]
No, I'm not sure it needs to be on the top level; I guess we just have to test it.

File: _header [hunk 1]
```diff
@@ -0,0 +1,10 @@
+Cache-Control: no-cache, no-store, max-age=0, must-revalidate
```
Comment (DaniRey) [inline _header]: Why do you want to configure no-cache headers?
We should always aim for the exact opposite.

File: _header [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+Pragma: no-cache
+Expires: 0
+X-Content-Type-Options: nosniff
+X-Frame-Options: DENY
```
Comment (DaniRey) [inline _header]: I don't see a reason to prevent this page from being included in an iFrame.

File: _header [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+X-XSS-Protection: 1; mode=block
+referrer-policy: no-referrer, strict-origin-when-cross-origin
+Strict-Transport-Security: max-age=31536000 ; includeSubDomains
+content-security-policy: default-src 'none'; script-src 'self' 'unsafe-inline' ajax.googleapis.com cdn.syndication.twimg.com; connect-src 'self'; img-src 'self' data: pbs.twimg.com *.travis-ci.org travis-ci.org; style-src 'self' 'unsafe-inline' fonts.googleapis.com; font-src 'self' data: fonts.gstatic.com; object-src 'none'; frame-ancestors 'none';report-uri https://unic.report-uri.com/r/d/csp/enforce; report-to csp
+Report-To: { "url": "https://unic.report-uri.com/r/d/csp/reportOnly”, "group": “cspro”, "max-age": 10886400 }, { "url": "https://unic.report-uri.com/r/d/csp/enforce”, "group": “csp","max-age": 10886400 }
```
Comment (DaniRey) [inline _header:10]: The max-age is very high. Do we have enough experience with this service, that we want to bind our website for more than half a year to this?

File: _header [hunk 4]
```diff
@@ -0,0 +1,10 @@
+Cache-Control: no-cache, no-store, max-age=0, must-revalidate
+Pragma: no-cache
+Expires: 0
+X-Content-Type-Options: nosniff
+X-Frame-Options: DENY
+X-XSS-Protection: 1; mode=block
+referrer-policy: no-referrer, strict-origin-when-cross-origin
+Strict-Transport-Security: max-age=31536000 ; includeSubDomains
```
Comment (DaniRey) [inline _header:8]: I don't want to include the subDomains just yet. We don't know if we want to use any of the subdomains for some special purpose which doesn't support HTTPS.

File: _header [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+content-security-policy: default-src 'none'; script-src 'self' 'unsafe-inline' ajax.googleapis.com cdn.syndication.twimg.com; connect-src 'self'; img-src 'self' data: pbs.twimg.com *.travis-ci.org travis-ci.org; style-src 'self' 'unsafe-inline' fonts.googleapis.com; font-src 'self' data: fonts.gstatic.com; object-src 'none'; frame-ancestors 'none';report-uri https://unic.report-uri.com/r/d/csp/enforce; report-to csp
```
Comment (DaniRey) [inline _header:9]: Where you able to verify that this content-security-policy is complete?

File: _header (hunk 2 above)
Comment (marcel-wiedemeier) [reply _header]: This is a basic security measure. When designing for security, you should not ask for reasons to make something secure, but for reasons why it should be possible to frame this site. There might be legitimate reasons for doing so, but even then you want to control the sites being allowed to do so. In this case, X-Frame-Options: DENY is a defense against clickjacking (https://www.owasp.org/index.php/Clickjacking_Defense_Cheat_Sheet), addressing A1:2017- Injection in the OWASP Top 10 2017.

File: _header (hunk 5 above)
Comment (marcel-wiedemeier) [reply _header:9]: I tested the policy according to https://projects.unic.com/x/G9Px. If it's not complete we'll get reports and can add any missing items.

File: _header (hunk 4 above)
Comment (marcel-wiedemeier) [reply _header:8]: I can't think of any purpose on our modern web not supporting HTTPS -especially as Chrome will mark non-HTTPS sites as not secure and certificates are free. Google has even includeSubDomains on some TLDs: https://security.googleblog.com/2017/09/broadening-hsts-to-secure-more-of-web.html

File: _header (hunk 3 above)
Comment (marcel-wiedemeier) [reply _header:10]: No issue as the reporting end point can be removed at any time by setting the value to 0:
 3.1.4. The max-age member
The REQUIRED max-age member defines the reporting endpoint’s lifetime, as a non-negative integer number of seconds. The member’s value MUST be a number; any other type will result in a parse error.

A value of "0" will cause the endpoint to be removed from the user agent’s reporting cache.

https://www.w3.org/TR/reporting-1/

File: _header (hunk 1 above)
Comment (marcel-wiedemeier) [reply _header]: I disagree and so does OWASP: https://www.owasp.org/index.php/OWASP_Application_Security_FAQ#Browser_Cache. Cache should be set deliberately set on objects than should be cached. The default should be off.

File: _header (hunk 3 above)
Comment (DaniRey) [reply _header:10]: Being able to actively flash the reporting cache is nice. Didn't know that. All good then.

File: _header (hunk 2 above)
Comment (DaniRey) [reply _header]: As this is an open-source project (with Apache license) I believe people should be able to use it in any ways they please. And I'm totally ok with it being framed in another website.
As it is a static webpage, where you cannot trigger any dangerous actions, I believe this security measure is not needed. Please remove this header.

File: _header (hunk 4 above)
Comment (DaniRey) [reply _header:8]: You convinced me. Open-source projects should always be leading examples.

File: _header (hunk 5 above)
Comment (DaniRey) [reply _header:9]: Perfect, thank you!

File: _header (hunk 1 above)
Comment (DaniRey) [reply _header]: Of course. My response was too generic to be true. We should always aim for caching, but of course rather cache too few then too much/long.

As this is a static website I would go for 
`Cache-Control: public, max-age=3600`

I'm not aware of a more fine granulare way to control cache headers with netlify. If so we could of course go for this as a standard and then define more specific cache headers for each resource type (html, css, js, ...)

Comment (marcel-wiedemeier) [reply _header]: Done.

File: _header (hunk 2 above)
Comment (DaniRey) [reply _header]: As we found at least one possible attack vector. Click-jacking the download and providing malicious content, I'm convinced that even for a static website `X-Frame-Options: DENY` is the way to go. Thank you for keeping to it.

**Pull request merged**